tests/test_Sequence.cc

```diff
@@ -3911,3 +3911,77 @@ TEST_CASE("Sequence: Check setter/getter function for step setup script", "[Sequ
         REQUIRE(ctx.step_setup_script == script);
     }
 }
+
+TEST_CASE("Sequence: sequence timeout", "[Sequence]")
+{
+    Step step{Step::type_action};
+    step.set_script("sleep(0.2)");
+
+    Sequence seq{"test_sequence"};
+    seq.push_back(std::move(step));
+
+    seq.set_timeout(100ms);
+    REQUIRE(seq.get_timeout() == 100ms);
+
+    Context ctx;
+    REQUIRE(seq.get_time_of_last_execution().time_since_epoch().count() == 0L);
+    auto maybe_error = seq.execute(ctx, nullptr);
+    REQUIRE(seq.get_time_of_last_execution().time_since_epoch().count() != 0L);
+
+    REQUIRE(maybe_error.has_value() == true);
+    REQUIRE_THAT(maybe_error->what(), Contains("Timeout: Sequence"));
+}
+
+TEST_CASE("Sequence: sequence timeout is lower than step timeout", "[Sequence]")
+{
+    Step step_1{Step::type_action};
+    step_1.set_script("sleep(600)"); // time constraint: set to a very long sleep of 10min
+
+    Sequence seq{"test_sequence"};
+    seq.push_back(std::move(step_1));
+    seq.set_timeout(500ms); // give the sequence enough time to execute the first step
+
+    Context ctx;
+    auto start = Clock::now();
+    auto maybe_error = seq.execute(ctx, nullptr);
+    auto end = Clock::now();
+
+    // check if executing step (sleeps 10min!) is lower then sequence timeout (500ms+100ms)
+    REQUIRE((end-start).count() < TimePoint(500ms + 100ms).time_since_epoch().count());
+}
+
+TEST_CASE("Sequence: test parallel sequences with timeout", "[Sequence]")
+{
+    Step step_1_1{Step::type_action};
+    step_1_1.set_script("sleep(500)");
+
+    Sequence seq_1{"test_sequence_1"};
+    seq_1.push_back(std::move(step_1_1));
+    seq_1.set_timeout(500ms); // sequence 1: timeout of 500ms
+
+    Step step_2_1{Step::type_action};
+    step_2_1.set_script("sleep(500)");
+
+    Sequence seq_2{"test_sequence_2"};
+    seq_2.push_back(std::move(step_2_1));
+    seq_2.set_timeout(500ms); // sequence 2: timeout of 500ms
+
+    Context ctx_1;
+    Context ctx_2;
+
+    std::thread t1([&seq_1, &ctx_1]() {
+        seq_1.set_timeout(std::move(Timeout{250ms})); // sequence 1: reset to 250ms!
+        auto maybe_error = seq_1.execute(ctx_1, nullptr);
+    });
+    std::thread t2([&seq_2, &ctx_2]() {
+        auto maybe_error = seq_2.execute(ctx_2, nullptr);
+    });
+
+    gul14::sleep(350ms);
+
+    REQUIRE(seq_1.is_timeout_elapsed() == true);
+    REQUIRE(seq_2.is_timeout_elapsed() == false);
```

Comment:
This is a race condition. There is no synchronization anywhere to make sure that you can modify `seq_1` and `seq_2` in one thread and simultaneously read them from another thread.

Reply:
Hmm, are you sure? Because both threads execute different sequences without a connection. I thought that a race condition manifest when two threads are executed simultaneously and accessing the same resource.

About the timing issue and as you already mentioned in some other remarks we cannot presume that the execution will possible stop for a while which has an influence on the elapsed time and can fail. 

I will remove the test case but since we have some 'controversy' about the race condition I wanted to keep it to have some remarks from you 😄 . Maybe I overlooked something...